### DISCUSSION TITLE
Lua utilities

### DISCUSSION THREAD

**Discussion by dndrks** [monome/iii#4] [Ideas]
thought it'd be useful to outline any utilities folks have pulled from norns for their iii sketches!

i've been using:
- [util.round](https://github.com/monome/norns/blob/e8e210ac4b7da320ac80c65ce605bbaccb6e7db8/lua/lib/util.lua#L191-L200)
- [util.clamp](https://github.com/monome/norns/blob/e8e210ac4b7da320ac80c65ce605bbaccb6e7db8/lua/lib/util.lua#L111-L118)
- [util.linlin](https://github.com/monome/norns/blob/e8e210ac4b7da320ac80c65ce605bbaccb6e7db8/lua/lib/util.lua#L140-L155)
- [util.wrap](https://github.com/monome/norns/blob/e8e210ac4b7da320ac80c65ce605bbaccb6e7db8/lua/lib/util.lua#L249-L266)

**Comment by tehn** [comment]
yes indeed all of those should be added, will get them in the next build!

**Comment by tehn** [comment]
trent suggested the `quote` lib from crow as a table serializer

https://github.com/monome/norns/blob/main/lua/core/crow/quote.lua

**Comment by tehn** [comment]
`250108` adds

- linlin
- wrap
- round
- clamp

**Comment by tehn** [comment]
still working on a table-to-string lib paired with some helpers for storing to flash.

also extending the midi helpers.

open to other additions!

**Comment by tehn** [comment]
upcoming:

- musicutil (streamlined)
- slew

**Comment by dndrks** [comment]
mm! is there any benefit to adding a helper to print booleans (for debugging)? i know we can use `ps("%d", 1==3)` or `print(tostring(1==3))`, but both feel like an extra move for a frequent case. eg. `pb(1==3)` feels fun!

**Reply by tehn** [reply]
yes absolutely. on my list is to improve `print` in general to do a lot more things that aren't in place, boolean included

also want `diii` to assume that if you leave a variable on the command line that it should be printed.

ie

```
x
```

infers `print(x)` like the standard lua repl

**Comment by justmat** [comment]
how would we feel about including the ``pattern_time`` lib?

**Reply by tehn** [reply]
yes absolutely want to rework this lib and include a perhaps better version than the original!

**Comment by dndrks** [comment]
having arc sensitivity baked into iii would be rad -- i've worked [the one from `encoders.lua`](https://github.com/monome/norns/blob/110ab7b98dc332df8adac2a35ffb526ab0408f9b/lua/core/encoders.lua#L53-L60) into the coming [SuperCollider `MonomeArc` class](https://github.com/monome/monomeSC/blob/c996977a60643c71b9966a882f49697afa4b593e/Classes/MonomeArc.sc#L145-L152) and it's really nice to have an abstraction take care of this typical use-case.

**Comment by justmat** [comment]
i've been making lots of variations on sequencers with arc, and find myself grabbing the euclidean stuff from norns fairly often. how do we feel about including it?

**Reply by tehn** [reply]
i think that's a worthy candidate!

we'll have a new beta candidate soon which represents a whole ton of improvements.

**Comment by ether-aj** [comment]
+1 to see musicutil in the next release, if possible
